src/rgw/driver/rados/rgw_putobj_processor.cc

```diff
@@ -552,6 +552,72 @@ int MultipartObjectProcessor::complete(
   op_target.set_versioning_disabled(true);
   op_target.set_meta_placement_rule(&tail_placement_rule);
 
+  if (if_match || if_nomatch) {
```

Comment:
just below, `MultipartObjectProcessor::complete()` is calling `obj_op.write_meta()` which calls your `target->check_preconditions()`. to use that, i think we just need to assign `obj_op.meta.if_match` and `obj_op.meta.if_nomatch` here